design/ceph-mon-pv.md

```diff
@@ -0,0 +1,139 @@
+offline for maintenance or upgrades.
+
+There are two major components to this design. First, the cluster CRD schema is
+modified to accept a PVC template for defining the storage to be used by
+monitors. And second, monitor pod management will be handled by StatefulSets
+rather than Deployments.
+
+# Monitor storage
+
+The monitor specification in the CRD is updated to include a persistent volume
+claim template that is used to obtain PVs for the monitor data store.
+
+```go
+type MonSpec struct {
+	Count                int  `json:"count"`
+	PreferredCount       int  `json:"preferredCount"`
+	AllowMultiplePerNode bool `json:"allowMultiplePerNode"`
+	VolumeClaimTemplate  *v1.PersistentVolumeClaim
+}
+```
+
+Unlike the storage requirements of OSDs (xf: StorageClassDeviceSets) it is
+reasonable to use default requirements for monitors (e.g. 5-10 GB). This means
+that a minimal claim may be specified using only the requested storage class in
+which case default resource requests will be used.
+
+```yaml
+volumeClaimTemplates:
+- spec:
+    storageClassName: gp2-ebs
+...
+volumeClaimTemplates:
+- spec:
+    storageClassName: local-storage
+```
+
+It is also reasonable for monitors to use storage in the form of a directory on
+the host's root file system. This behavior effectively provides something like
+Rook-specific dynamic HostPath provisioning and is the current default behavior
+in Rook. In order to continue to support this level of operational simplicity,
+the volume claim template itself can be left unspecified entirely to request the
+existing default behavior. Note that the storage class `local-storage` is not
+necessarily equivalent---there are a variety of implementations with the most
+popular requiring static provisioning [0, 1].
+
+[0]: https://github.com/rancher/local-path-provisioner
+[1]: https://github.com/kubernetes-sigs/sig-storage-local-static-provisioner
+
+## Non-PV monitor storage
+
+Not all monitor storage requirements should be satisfied using PV storage. Most
+notably log and crash data. These data may continue to be stored in HostPath
+volume mounts. This will continue to be based on `dataDirHostPath`.
+
+## Host networking
+
+Currently Ceph monitor pods may be deployed using host-based networking, or
+placed behind a service using a portable network address. This poses a challenge
+when migrating a monitor (and it's PV) to a new node which has been deployed
+using host-based networking as the monitor's address will no longer valid.
+
+Support for migration of monitors with host-based networking is future work. To
+simplify the initial implementation, host-based networking will not be supported
+for non-local PVs.
+
+# StatefulSet pod management
+
+Monitor pods will be managed by a StatefulSet rather than Deployments which are
+currently use. This change offers (1) simplified naming for DNS and pods (e.g.
+`ceph-mon-a` rather than `ceph-mon-a-3ab423fb`), and (2) StatefulSets are able
```

Comment:
The pod name will still have a numerical suffix such as `ceph-mon-a-0`

Reply:
yes